godot-core/src/obj/traits.rs

```diff
@@ -137,6 +139,16 @@ pub mod cap {
         fn __godot_init(base: Base<Self::Base>) -> Self;
     }
 
+    pub trait GodotToString: GodotClass {
+        #[doc(hidden)]
+        fn __virtual_to_string(&self) -> GodotString;
+    }
+
+    pub trait GodotRegisterClass: GodotClass {
+        #[doc(hidden)]
+        fn __virtual_register_class(builder: &mut ClassBuilder<Self>);
+    }
```

Comment:
Do these need to be public traits? (Note that "public" doesn't necessarily mean `pub`; as you see with many symbols, `#[doc(hidden)]` allows them to be used in proc-macros without making them part of the public APIs).

`GodotInit` is public because it's used as a generic bound, to determine whether an object is default-constructible from Godot or not. I don't think `to_string` and `register_class` contribute to the type in such a way (maybe this could be discussed for `GodotToString`, similar to `Display`). I think they are more like `ImplementsGodotVirtual` below. However, I see that you use them as bounds in the builder API -- possibly something to revisit once we get to that.

One issue with the current approach is that every time we add another method to what was previously in `GodotExt`, we need to code-generate a new trait and adjust the bounds in several places. But I'm also fine to leave it for now and address that problem, once we have it.

---

What I would definitely change is to remove the "virtual" from the method name, as those method are not technically related to Godot virtuals. The 2 underscores prefix can stay.

Reply:
My logic here was to make these other traits like `GodotInit`, since they follow the same principles: Can be used as a Rust-side bound, intended to be implemented automatically based on `#[godot_api]`, and is a special case in Godot to begin with. Open to other suggestions for how to handle that though. Will remove `__virtual`.

Reply:
Regarding `__register_class`, I don't know if this abstraction would be necessary in user-facing code? No one is ever going to invoke the class registration method manually, and I also don't think whether a class has a registration method or not should be part of its public API.

`__to_string` is more reasonable, but also here, I'm not sure if the Rust-side bound fully reflects reality: in Godot, it's possible to invoke `to_string()` on any object, independently of whether it implements such a method or not. So we may constrain implementations more than necessary by requiring such a bound.

Maybe you could mark them `#[doc(hidden)]` for now and add a `// TODO` -- we probably have to revisit this anyway once we get to builder APIs. Until then, I don't really see a use case for having these traits public.